gdsfactory/functions.py

```diff
@@ -498,3 +498,62 @@ def mirror(component: Component, x_mirror: bool = True) -> gf.Component:
     c.add_ports(ref.ports)
     c.copy_child_info(component)
     return c
+
+
+def remove_shapes_near_exclusion(
+    c: gf.Component,
+    target_layer: LayerSpec,
+    exclusion_layer: LayerSpec,
+    *,
+    margin_um: float = 2.0,
+    remove_entire_shapes: bool = True,
+    flatten: bool = True,
+) -> gf.Component:
+    """Remove shapes on target_layer that interact with exclusion_layer.
+
+    Args:
+        c: Component to modify.
+        target_layer: Layer containing shapes to potentially remove.
+        exclusion_layer: Layer defining exclusion zones.
+        margin_um: Exclusion margin/halo in microns (default 2.0).
+        remove_entire_shapes: If True, removes entire shapes that touch the
+            exclusion zone. If False, only clips the overlapping portions.
+        flatten: If True, flattens the component before processing.
+
+    Returns:
+        Modified component with shapes removed/clipped.
+    """
+    import klayout.db as kdb
+
+    if flatten:
+        c.flatten()
+
+    target_layer_idx = gf.get_layer(target_layer)
+    exclusion_layer_idx = gf.get_layer(exclusion_layer)
+
+    # Convert margin to database units
+    dbu = c.kcl.dbu
+    margin_dbu = int(margin_um / dbu)
+
+    # Get the exclusion region and expand it
+    exclusion_layer_kdb = c.kcl.layer(*exclusion_layer_idx)
+    exclusion_region = kdb.Region(c.begin_shapes_rec(exclusion_layer_kdb))
+    halo_region = exclusion_region.sized(margin_dbu)
+
+    # Get target shapes
+    target_layer_kdb = c.kcl.layer(*target_layer_idx)
+    target_region = kdb.Region(c.begin_shapes_rec(target_layer_kdb))
+
+    if remove_entire_shapes:
+        # Remove entire shapes that interact with the exclusion halo
+        # A shape "interacts" if it has any overlap with the halo
+        interacting = target_region.interacting(halo_region)
```

Comment:
I would suggest to use `overlapping` instead of interacting here. Otherwise you grab stuff that is exactly at the minimum of the rule as well.